sync/class.jetpack-sync-module-import.php

```diff
@@ -2,86 +2,148 @@
-		 *
-		 * @since 5.0.0
-		 *
-		 * $param string $importer
-		 */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
-		$this->import_end = true;
-	}
+	public function sync_import_action( $importer ) {
+		$import_action = current_filter();
+		// map action to event name
+		$sync_action = self::$import_sync_action_map[ $import_action ];
 
-	public function sync_import_end() {
-		// We already ran an send the import
-		if ( $this->import_end ) {
+		// Only sync each action once per import
+		if ( array_key_exists( $sync_action, $this->synced_actions ) && $this->synced_actions[ $sync_action ] ) {
 			return;
 		}
 
-		$this->import_end = true;
-		$importer         = 'unknown';
-		$backtrace        = wp_debug_backtrace_summary( null, 0, false );
-		if ( $this->is_importer( $backtrace, 'Blogger_Importer' ) ) {
-			$importer = 'blogger';
-		}
+		// Mark this action as synced
+		$this->synced_actions[ $sync_action ] = true;
 
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WC_Tax_Rate_Importer' ) ) {
-			$importer = 'woo-tax-rate';
+		// prefer self-reported $importer value
+		if ( ! $importer ) {
+			// fall back to inferring by calling class name
+			$importer = self::get_calling_importer_class();
 		}
 
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WP_Import' ) ) {
-			// phpcs:ignore WordPress.WP.CapitalPDangit
-			$importer = 'wordpress';
+		// Get $importer from known_importers
+		if ( isset( self::$known_importers[ $importer ] ) ) {
+			$importer = self::$known_importers[ $importer ];
 		}
 
 		$importer_name = $this->get_importer_name( $importer );
 
-		/** This filter is already documented in sync/class.jetpack-sync-module-posts.php */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
+		do_action( $sync_action, $importer, $importer_name );
 	}
 
 	private function get_importer_name( $importer ) {
 		$importers = get_importers();
 		return isset( $importers[ $importer ] ) ? $importers[ $importer ][0] : 'Unknown Importer';
 	}
 
-	private function is_importer( $backtrace, $class_name ) {
-		foreach ( $backtrace as $trace ) {
-			if ( strpos( $trace, $class_name ) !== false ) {
-				return true;
+	/**
+	 * Determine the class that extends `WP_Importer` which is responsible for
+	 * the current action. Designed to be used within an action handler.
+	 *
+	 * @return string  The name of the calling class, or 'unknown'.
+	 */
+	private static function get_calling_importer_class() {
+		// If WP_Importer doesn't exist, neither will any importer that extends it
+		if ( ! class_exists( 'WP_Importer' ) ){
+			return 'unknown';
+		}
+
+		$action = current_filter();
+		$backtrace = wp_debug_backtrace_summary( null, 0, false );
+
+		$do_action_pos = -1;
+		for ( $i = 0; $i < count( $backtrace ); $i++ ) {
+			// Find the location in the stack of the calling action
+			if ( preg_match( "/^do_action\\(\'([^\']+)/", $backtrace[ $i ], $matches ) ) {
```

Comment:
Is it worth guarding against implementation changes in `wp_debug_backtrace_summary`, etc. by allowing for optional whitespace around the argument in `do_action`?

Something like this:

```suggestion
			if ( preg_match( "/^do_action\s*\\(\s*\'([^\']+\s*)/", $backtrace[ $i ], $matches ) ) {
```

Reply:
Thanks, great idea -- I just committed your suggestion!